compiler/rustc_passes/src/stability.rs

```diff
@@ -805,14 +806,69 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
     }
 
     fn visit_path(&mut self, path: &'tcx hir::Path<'tcx>, id: hir::HirId) {
-        if let Some(def_id) = path.res.opt_def_id() {
-            let method_span = path.segments.last().map(|s| s.ident.span);
-            self.tcx.check_stability(def_id, Some(id), path.span, method_span)
+        if !is_unstable_reexport(self.tcx, path, id) {
+            if let Some(def_id) = path.res.opt_def_id() {
+                let method_span = path.segments.last().map(|s| s.ident.span);
+                self.tcx.check_stability(def_id, Some(id), path.span, method_span)
+            }
         }
         intravisit::walk_path(self, path)
     }
 }
 
+/// Check whether the path is a re-export of an unstable item, that has also been marked as unstable under the same feature name.
+fn is_unstable_reexport<'a, 'tcx>(
+    tcx: TyCtxt<'tcx>,
+    path: &'a hir::Path<'tcx>,
+    id: hir::HirId,
+) -> bool {
+    // Get the LocalDefId so we can lookup the item to check the kind.
+    let def_id = if let Some(def_id) = tcx.hir().opt_local_def_id(id) {
+        def_id
+    } else {
+        return false;
+    };
+
+    let stab = if let Some(stab) = tcx.stability().local_stability(def_id) {
+        stab
+    } else {
+        return false;
+    };
+
+    let item = tcx.hir().item(hir::ItemId { def_id });
+    let item_kind = &item.kind;
+
+    // If this is a path that isn't a use, we don't need to do anything special
+    if !matches!(item_kind, ItemKind::Use(..)) {
+        return false;
+    }
+
+    let decision = if let Some(def_id) = path.res.opt_def_id() {
+        let method_span = path.segments.last().map(|s| s.ident.span);
+
+        tcx.eval_stability(def_id, Some(id), path.span, method_span)
+    } else {
+        // If the path doesn't resolve to anything, there's nothing more to check.
+        return false;
+    };
+
+    // If we are going to deny the usage, check whether it's a re-export first with the same unstable feature.
+    if let rustc_middle::middle::stability::EvalResult::Deny { feature, .. } = decision {
```

Comment:
I don't think this condition is necessary?
It doesn't matter whether the reexport target is unstable or not, and whether the feature is enabled or disabled, the reexport gating can be skipped anyway.

Reply:
I went ahead and removed these checks. 

As it is now this PR would completely prevent any `use` items from triggering a "use of unstable" error, only the actual usages of the item trigger errors. Do you think checking whether staged_api is enabled would be good?